llvm/lib/Analysis/MemoryProfileInfo.cpp

```diff
@@ -301,6 +316,22 @@ bool CallStackTrie::buildAndAttachMIBMetadata(CallBase *CI) {
                                 "single");
     return false;
   }
+  // If there were any hot allocation contexts, the Alloc trie node would have
+  // the Hot type set. If so, because we don't currently support cloning for hot
+  // contexts, they should be converted to NotCold. This happens in the cloning
+  // support anyway, however, doing this now enables more aggressive context
+  // trimming when building the MIB metadata (and possibly may make the
+  // allocation have a single NotCold allocation type), greatly reducing
+  // overheads in bitcode, cloning memory and cloning time.
+  if (Alloc->AllocTypes & static_cast<uint8_t>(AllocationType::Hot)) {
+    convertHotToNotCold(Alloc);
```

Comment:
I'm not clear on why we need this. If we set `MemProfUseHotHints` to false, then no alloctypes should have hot type and we'll never enter this condition?

Reply:
By default hot hint generation is off, however, since it is still optionally supported this avoids a big overhead increase if that is re-enabled for experimentation etc. If we keep that support here, even if optional, I want to ensure things keep working the same.

Reply:
I see, so this change really does 2 things -- optimize when hot hints are used and allow disabling them altogether. The description now makes more sense to me. Could you commit this as two separate changes -- adding the flag and the optimization (conversion to notcold)? For the former it would be good to have a small test for the flag usage.

Reply:
ok I can do that. There is a unittest that tests with the flag, but I'll update llvm/test/Transforms/PGOProfile/memprof.ll to test with this as well

Reply:
See PR124338. I'll update this one to be just about the conversion once that goes in.

Reply:
Updated, ptal

Reply:
> For the former it would be good to have a small test for the flag usage.

I had to remove the test I added there as it checks metadata, and I had forgotten that with this change that would no longer happen. I will submit a follow on test that checks the option, using an unambiguously hot allocation.